components/formats-gpl/src/loci/formats/in/HarmonyReader.java

```diff
@@ -0,0 +1,783 @@
+    suffixSufficient = false;
+    hasCompanionFiles = true;
+    datasetDescription =
+      "Directory with XML file and one .tif/.tiff file per plane";
+  }
+
+  // -- IFormatReader API methods --
+
+  /* @see loci.formats.IFormatReader#getRequiredDirectories(String[]) */
+  @Override
+  public int getRequiredDirectories(String[] files)
+    throws FormatException, IOException
+  {
+    return 1;
+  }
+
+  /* @see loci.formats.IFormatReader#isSingleFile(String) */
+  @Override
+  public boolean isSingleFile(String id) throws FormatException, IOException {
+    return false;
+  }
+
+  /* @see loci.formats.IFormatReader#fileGroupOption(String) */
+  @Override
+  public int fileGroupOption(String id) throws FormatException, IOException {
+    return FormatTools.MUST_GROUP;
+  }
+
+  /* @see loci.formats.IFormatReader#isThisType(String, boolean) */
+  @Override
+  public boolean isThisType(String name, boolean open) {
+    String localName = new Location(name).getName();
+    boolean exists = false;
+    for (String XML_FILE : XML_FILES) {
+      if (localName.equals(XML_FILE)) {
+        exists = true;
+        break;
+      }
+    }
+    if (!exists) {
+      Location parent = new Location(name).getAbsoluteFile().getParentFile();
+      for (String XML_FILE : XML_FILES) {
+        Location xml = new Location(parent, XML_FILE);
+        if (xml.exists()) {
+          exists = true;
+          break;
+        }
+      }
+    }
+    if (!exists) {
+      return false;
+    }
+
+    return super.isThisType(name, open);
+  }
+
+  /* @see loci.formats.IFormatReader#isThisType(RandomAccessInputStream) */
+  @Override
+  public boolean isThisType(RandomAccessInputStream stream) throws IOException {
+    TiffParser p = new TiffParser(stream);
+    IFD ifd = p.getFirstIFD();
+    if (ifd != null) {
+      Object s = ifd.getIFDValue(XML_TAG);
+      if (s == null) return false;
+      String xml = s instanceof String[] ? ((String[]) s)[0] : s.toString();
+      return xml.indexOf(MAGIC) < 1024;
+    }
+    stream.seek(0);
+    String xml = stream.readString(1024);
+    return xml.indexOf(MAGIC) > 0;
+  }
+
+  /* @see loci.formats.IFormatReader#getUsedFiles(boolean) */
+  @Override
+  public String[] getUsedFiles(boolean noPixels) {
+    FormatTools.assertId(currentId, true, 1);
+
+    ArrayList<String> files = new ArrayList<String>();
+    files.addAll(metadataFiles);
+    if (!noPixels) {
+      for (Plane[] well : planes) {
+        for (Plane p : well) {
+          files.add(p.filename);
+        }
+      }
+    }
+    return files.toArray(new String[files.size()]);
+  }
+
+  /* @see loci.formats.IFormatReader#getSeriesUsedFiles(boolean) */
+  @Override
+  public String[] getSeriesUsedFiles(boolean noPixels) {
+    FormatTools.assertId(currentId, true, 1);
+
+    ArrayList<String> files = new ArrayList<String>();
+    files.addAll(metadataFiles);
+    if (!noPixels) {
+      for (Plane p : planes[getSeries()]) {
```

Comment:
Just because you do this below... doesn't this required a check getSeries() < planes.length?